cli/tests/test_abandon_command.rs

```diff
@@ -176,6 +176,193 @@ fn test_basics() {
+    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
+    @  [znk] c
+    ◉  [roy] a b??
+    ◉  [zsu] b?? base
+    ◉  [rlv] nottherootcommit
+    ◉  [zzz]
+    "###);
+
+    test_env.jj_cmd_ok(&repo_path, &["op", "restore", &setup_opid]);
+    // ========= Reminder of the setup ===========
+    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
+    @  [znk] c
+    ◉    [vru] b
+    ├─╮
+    │ ◉  [roy] a
+    ├─╯
+    ◉  [zsu] base
+    ◉  [rlv] nottherootcommit
+    ◉  [zzz]
+    "###);
+    let (stdout, stderr) = test_env.jj_cmd_ok(&repo_path, &["abandon", "a", "b"]);
+    insta::assert_snapshot!(stdout, @"");
+    insta::assert_snapshot!(stderr, @r###"
+    Abandoned the following commits:
+      royxmykx 98f3b9ba a | a
+      vruxwmqv 8c0dced0 b | b
+    Rebased 1 descendant commits onto parents of abandoned commits
+    Working copy now at: znkkpsqq 84fac1f8 c | c
+    Parent commit      : zsuskuln 73c929fc a b base | base
+    Added 0 files, modified 0 files, removed 2 files
+    "###);
+    // This is likely what the user would expect
+    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
+    @  [znk] c
+    ◉  [zsu] a b base
+    ◉  [rlv] nottherootcommit
+    ◉  [zzz]
+    "###);
+    let (stdout, stderr) = test_env.jj_cmd_ok(&repo_path, &["branch", "list", "b"]);
+    insta::assert_snapshot!(stdout, @r###"
+    b: zsuskuln 73c929fc base
+    "###);
+    insta::assert_snapshot!(stderr, @"");
+}
+
+#[test]
+fn test_bug_2600_rootcommit_special_case() {
+    let test_env = TestEnvironment::default();
+    test_env.jj_cmd_ok(test_env.env_root(), &["init", "repo", "--git"]);
+    let repo_path = test_env.env_root().join("repo");
+
+    // Set up like `test_bug_2600`, but without the `nottherootcommit` commit.
+    create_commit(&test_env, &repo_path, "base", &[]);
+    create_commit(&test_env, &repo_path, "a", &["base"]);
+    create_commit(&test_env, &repo_path, "b", &["base", "a"]);
+    create_commit(&test_env, &repo_path, "c", &["b"]);
+
+    // Setup
+    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
+    @  [vru] c
+    ◉    [roy] b
+    ├─╮
+    │ ◉  [zsu] a
+    ├─╯
+    ◉  [rlv] base
+    ◉  [zzz]
+    "###);
+
+    // Now, the test
+    let (stdout, stderr) = test_env.jj_cmd_ok(&repo_path, &["abandon", "base"]);
+    insta::assert_snapshot!(stdout, @"");
+    insta::assert_snapshot!(stderr, @r###"
+    Abandoned commit rlvkpnrz 0c61db1b base | base
+    Rebased 3 descendant commits onto parents of abandoned commits
+    Working copy now at: vruxwmqv 73e9185c c | c
+    Parent commit      : royxmykx 80dd9cba b | b
+    Added 0 files, modified 0 files, removed 1 files
+    "###);
+    // The current behavior is either correct or should be replaced with an error
+    // message. Even though the user would expect `b` to still be a descendant of
+    // `base`, it is impossible in the Git backend.
```

Comment:
This use of "impossible" made me think again whether it's actually impossible :) And, actually, we could simply support storing overrides for the set of parents outside the Git commit (where we already store the change id and conflict data).

Reply:
Also see my suggestion on #2655 about moving these tests into the library crate.

Reply:
I was wondering about allowing this, we could simply store the list of children of the root commit. We could alternatively have a fake root commit in the repo, and have jj know that children of that commit actually represent children of the root commit that are merge commits. Then, we'd only need to store its id.

OTOH, it might be weird if this was the one exception to git parents being different from jj parents.

Re moving to the library, I replied in the other PR. TLDR, I think not yet.

Reply:
> I was wondering about allowing this, we could simply store the list of children of the root commit. We could alternatively have a fake root commit in the repo, and have jj know that children of that commit actually represent children of the root commit that are merge commits. Then, we'd only need to store its id.

I think it would be simpler to do it as I suggested above (storing it in the same storage as we already store change ids and conflict data).

> Re moving to the library, I replied in the other PR. TLDR, I think not yet.

Sure, no problem.

Reply:
> I think it would be simpler to do it as I suggested above (storing it in the same storage as we already store change ids and conflict data).

You're likely right. A fake commit has some minor UI advantages, though. It could have a nice description, so that Git users can recover the actual tree structure. It could also be pushed to a remote (which is both an advantage and a disadvantage; the description would need some marker JJ can recognize when pulling from remotes).

In fact, if we just store it with change ids, re-cloning a repo from Git will change it's commit graph. This wouldn't be the end of the world, but is still not great.